### DISCUSSION TITLE
x265: Fix Linux bug to utilize all the cores

### DISCUSSION THREAD

**Discussion by alatteri** [CESNET/UltraGrid#278] [Ideas]
I saw this commit on the x265 repo master.  It has been a long time since they cut a release and this fix is outside of the last 3.5 cut.  I'm wondering if it is worthwhile to integrate a compile based off master for the AppImage.

https://bitbucket.org/multicoreware/x265_git/commits/a7a97ab32848d7d24a2a6e0de8b39ad98906ae1a

**Comment by MartinPulec** [comment]
Hi, for this particular commit - I don't think it is worth including notably for this one:

1. it seems to me that it even doesn't have any effect - supposing that it enables [this if-branch](https://bitbucket.org/multicoreware/x265_git/src/a7a97ab32848d7d24a2a6e0de8b39ad98906ae1a/source/common/threadpool.cpp#lines-305), which sets NUMA configuration, let say "8,8" to "pools" option, which is transcribed back to `p->numaPools`. _But_, even if it is not run and the value is left default, which I believe is NULL, [this branch](https://bitbucket.org/multicoreware/x265_git/src/a7a97ab32848d7d24a2a6e0de8b39ad98906ae1a/source/common/threadpool.cpp#lines-368) does the same work.
2. it is NUMA specific. We discussed and we think that nowadays isn't much suitable for libx265 - FFmpeg complains even when using more than 16 threads, which I suppose is because it doesn't scale well for more. But today even single node CPU have 16 logical units (with or without SMT), so it would be definitely better to keep it on a single node.
3. if I am mistaken, and the point 1. doesn't hold, still the NUMA configuration can be passed explicitly as `-c libavcodec:x265-params=pool=8,8`

Anyways, while looking into that, I came across [some more options](https://x265.readthedocs.io/en/master/cli.html?highlight=numa#performance-options). If you want to increase parallelism, have you already tried some of `pmode`, `pme` or `slices` parameters? It should go into UG libavcodec `x265-params` option, eg. `'-c libavcodec:encoder=libx265:x265-params=pme=1\:pmode=1\:slices=10` gave me slight (20%) performance gain. But of course it depends, may also have no effect if the utilization is already high.

**Reply by alatteri** [reply]
Hi Martin,

Yeah, I think I've tried every combo option to try and get x265 4K realtime.  The best I can get is ~ 21fps on a AMD Ryzen 9 5950X, that is with a system load of around 25.  In additional, I have to specific threads=2, which at least with x265 really corresponds to frame-threads and not encoding threads. https://x265.readthedocs.io/en/stable/cli.html?highlight=frame-threads#cmdoption-frame-threads

```
x265 [info]: Thread pool created using 32 threads
x265 [info]: Slices                              : 10
x265 [info]: frame threads / pool features       : 2 / wpp(68 rows)+pmode+pme 
[lavcd aud.] Using audio encoder: aac
x265 [info]: Coding QT: max CU size, min CU size : 32 / 16
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : dia / 57 / 0 / 2
x265 [info]: Cb/Cr QP Offset                     : 6 / 6
x265 [info]: Lookahead / bframes / badapt        : 0 / 0 / 0
x265 [info]: b-pyramid / weightp / weightb       : 0 / 0 / 0
x265 [info]: References / ref-limit  cu / depth  : 1 / off / off
x265 [info]: Rate Control / qCompress            : CRF-20.0 / 0.50
x265 [info]: tools: rd=2 psy-rd=2.00 early-skip rskip mode=1 tmvp cip
x265 [info]: tools: fast-intra strong-intra-smoothing lslices=8 slices=10
x265 [info]: tools: deblock
```

```
[DeckLink capture] 105 frames in 5.03924 seconds = 20.8365 FPS
[Audio sender] Sent 240000 samples in last 5.012529 seconds.
[Audio sender] Volume: [0] -28.45/-12.72, [1] -29.37/-12.89, [2] -30.43/-8.77, [3] -90.78/-68.73, [4] -30.92/-15.23, [5] -32.01/-15.43, [6] -24.60/-8.38, [7] -25.37/-9.37 dBFS RMS/peak
[DeckLink capture] 106 frames in 5.01345 seconds = 21.1431 FPS
[DeckLink capture] 104 frames in 5.00884 seconds = 20.7633 FPS
[Audio sender] Sent 244000 samples in last 5.056084 seconds.
[Audio sender] Volume: [0] -30.38/-10.46, [1] -31.45/-11.31, [2] -25.56/-2.29, [3] -46.44/-18.84, [4] -32.87/-13.00, [5] -34.15/-14.03, [6] -25.46/-5.28, [7] -26.04/-5.83 dBFS RMS/peak
[DeckLink capture] 105 frames in 5.01069 seconds = 20.9552 FPS
[Audio sender] Sent 240000 samples in last 5.011937 seconds.
[Audio sender] Volume: [0] -18.88/-0.24, [1] -18.90/-0.35, [2] -26.39/-7.38, [3] -29.72/-16.78, [4] -32.60/-13.43, [5] -33.92/-13.16, [6] -22.04/-1.85, [7] -22.14/-2.16 dBFS RMS/peak
[DeckLink capture] 106 frames in 5.04615 seconds = 21.0061 FPS
[Audio sender] Sent 242000 samples in last 5.044858 seconds.
[Audio sender] Volume: [0] -25.03/-9.84, [1] -25.20/-10.83, [2] -24.02/-0.50, [3] -28.90/-8.04, [4] -29.19/-7.27, [5] -30.01/-9.59, [6] -23.79/-5.85, [7] -24.04/-5.83 dBFS RMS/peak
[DeckLink capture] 103 frames in 5.02729 seconds = 20.4882 FPS
```


![Screen Shot 2023-01-10 at 3 29 36 PM](https://user-images.githubusercontent.com/625982/211683723-b6e868ba-d9d0-419a-a3ae-0991a794a002.png)
